test/smtcallback.js

```diff
@@ -7,88 +7,198 @@ const solc = require('../index.js');
+function collectFiles (testdir) {
+  let sources = [];
+  // BFS to get all test files
+  let dirs = [testdir];
+  while (dirs.length > 0) {
+    const dir = dirs.shift();
+    const files = fs.readdirSync(dir);
+    for (let i in files) {
+      const file = path.join(dir, files[i]);
+      if (fs.statSync(file).isDirectory()) {
+        dirs.push(file);
+      } else {
+        sources.push(file);
+      }
+    }
   }
+  return sources;
+}
+
+function createTests (sources, st) {
+  // Read tests and collect expectations
+  let tests = {};
+  for (let i in sources) {
+    st.comment('Collecting ' + sources[i] + '...');
+    const source = fs.readFileSync(sources[i], 'utf8');
+
+    let engine;
+    const option = '// SMTEngine: ';
+    if (source.includes(option)) {
+      let idx = source.indexOf(option);
+      if (source.indexOf(option, idx + 1) !== -1) {
+        st.skip('SMTEngine option given multiple times.');
+        continue;
+      }
+      const re = new RegExp(option + '(\\w+)');
+      let m = source.match(re);
+      assert(m !== undefined);
+      assert(m.length >= 2);
+      engine = m[1];
+    }
 
-  for (var i in errors) {
-    if (errors[i].includes('Error trying to invoke SMT solver') || expectations[i].includes('Error trying to invoke SMT solver')) {
+    if (source.includes('==== Source')) {
+      st.skip('Test requires soltest in source imports.');
       continue;
     }
-    // Expectations containing counterexamples might have many '\n' in a single line.
-    // These are stored escaped in the test format (as '\\n'), whereas the actual error from the compiler has '\n'.
-    // Therefore we need to replace '\\n' by '\n' in the expectations.
-    // Function `replace` only replaces the first occurrence, and `replaceAll` is not standard yet.
-    // Replace all '\\n' by '\n' via split & join.
-    expectations[i] = expectations[i].split('\\n').join('\n');
-    if (ignoreCex) {
-      expectations[i] = expectations[i].split('\nCounterexample')[0];
-      errors[i] = errors[i].split('\nCounterexample')[0];
+
+    let expected = [];
+    const delimiter = '// ----';
+    if (source.includes(delimiter)) {
+      expected = source.substring(source.indexOf('// ----') + 8, source.length).split('\n');
+      expected = expected.filter(line => isSMTCheckerErrorMessage(line));
+      expected = expected.map(line => errorFromExpectation(line));
     }
-    // `expectations` have "// Warning ... " before the actual message,
-    // whereas `errors` have only the message.
-    if (!expectations[i].includes(errors[i])) {
+    tests[sources[i]] = {
+      expectations: expected,
+      solidity: { test: { content: source } },
+      engine: engine
+    };
+  }
+  return tests;
+}
+
+function collectStringErrors (solOutput) {
+  assert(solOutput !== undefined);
+  return solOutput.errors.filter(error => !error.message.includes('This is a pre-release')).map(error => error.message);
+}
+
+function expectStringErrors (a, b) {
```

Comment:
These two functions have the same behavior as before and are only used by the first test.